### PR TITLE
Move autograd.Function.apply wrapper into C++

### PR DESCRIPTION
**Pull request by zou3519**
Stack from [ghstack](https://github.com/ezyang/ghstack/tree/0.15.0) (oldest at bottom):
* #190695
* __->__ #190694

Route torch.autograd.Function.apply directly to _C._FunctionBase.apply and move the Python wrapper work into THPFunction_apply. The C++ path now handles setup_context default-argument binding, functorch transform dispatch, dead-wrapper unwrapping, and kwargs normalization before continuing through the existing apply implementation.

On the local one-apply 13-in-2-out no-save instruction-count benchmark, this moved from 46,185 to 40,262 instructions, a reduction of 5,923 instructions (12.8%).

NB: This PR is now applying unwrap_dead_wrapers to all args
(args+kwargs) instead of all args. We do this because it was easier to
write code this way and this is technically a bugfix but cool that no
one actually ever ran into it.

Authored with assistance from an AI assistant.

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @kadeng @chauhang @amjames @jataylo @azahed98

### CODE REVIEW

File: torch/csrc/autograd/python_function.cpp [hunk 1]
```diff
@@ -1409,6 +1410,67 @@ static PyObject* get_base_setup_context() {
...
+  }
+  auto orig_setup_context = get_base_setup_context();
+  if (!orig_setup_context) {
+    return -1;
+  }
+  return cls_setup_context.get() != orig_setup_context;
+}
+
+static PyObject* get_autograd_function_attr(const char* name) {
+  auto module = THPObjectPtr(PyImport_ImportModule("torch.autograd.function"));
```
Comment (Skylion007) [inline torch/csrc/autograd/python_function.cpp]: prefer module_, module will be a keyword in CPP23

Comment (zou3519) [reply torch/csrc/autograd/python_function.cpp]: yup will fix

File: torch/autograd/function.py [hunk 2]
```diff
@@ -606,37 +605,24 @@ def vmap(info, in_dims, *args):
...
 
-        is_setup_ctx_defined = _is_setup_context_defined(cls.setup_context)
-        if is_setup_ctx_defined:
-            args, kwargs = bind_default_args(cls.forward, *args, **kwargs)
 
-        if not torch._C._are_functorch_transforms_active():
-            # See NOTE: [functorch vjp and autograd interaction]
-            args = _functorch.utils.unwrap_dead_wrappers(args)
-            return super().apply(*args, **kwargs)  # type: ignore[misc]
+def _bind_default_args(func, args, kwargs):
```
Comment (Skylion007) [inline torch/autograd/function.py]: Type this, arg and kwargs should be derived from the ParamSpec of func?

**Review by soulitzer — APPROVED** [review]

**Comment by zou3519** [comment]
## mergedog handoff

All CI is green (or skipped). Ready for human review and `@pytorchbot merge`.

Current PR head: `938e9a3417277b9d317f5c37b91d2314e24816ea`.

### Autonomous changes pushed

- `9d6e2cb836bd` — pushed fix commit 9d6e2cb836bd — failing jobs: linux-jammy-cuda13.0-py3.10-gcc11 / test (pr_time_benchmarks, 1, 1, mt-l-bx86iavx512-94-344-t4-8) (fix-CI (2026-07-23T03:06:44+00:00))

During shepherding, the configured LLM was invoked **2** times. If any CI shows red, it judged that failure unrelated to this PR's changes — please verify below before merging.

### Session 1 — fix-CI (2026-07-23T02:17:55+00:00)

- **Before:** `938e9a341727`
- **Verdict:** judged failures spurious (no commit) — failing jobs: bc_linter; rationale: BC Lint / bc_linter is an expected false positive already covered by the suppress-bc-linter label. Function.apply was not removed from the runtime API: it is now assigned directly from _C._FunctionBase, which the source-based BC linter cannot recognize as preserving the method.

<details><summary>LLM transcript</summary>

```
```

</details>

### Session 2 — fix-CI (2026-07-23T03:06:44+00:00)

- **Before:** `938e9a341727`
- **After:** `9d6e2cb836bd`
- **Verdict:** pushed fix commit 9d6e2cb836bd — failing jobs: linux-jammy-cuda13.0-py3.10-gcc11 / test (pr_time_benchmarks, 1, 1, mt-l-bx86iavx512-94-344-t4-8)

<details><summary>LLM transcript</summary>

```
```

</details>
